biocrnpyler/dna_construct.py

```diff
@@ -20,13 +20,284 @@
+                        #this happens if we do in fact make a protein
+                        proteins_per_rbs+=[protein_part[0]]
+            #print(rna_partslist)
+            #print("currently we are translating from "+str(rbs)+ " which is located at " + str(rbs.pos))
+            #print(rbs)
+            #print(rna_partslist)
+            if(self.return_rnas):
+                #this means we are creating new RNAs, and so the correct RBS is the one in the new RNA.
+                correct_rbs = rna_construct.parts_list[rna_partslist.index([rbs,"forward"])]
+            else:
+                #this means we are reading an existing RNA, so then the correct rbs is the one
+                #in the original "DNA", which is actually an RNA.
+                correct_rbs = rbs
+            proteins_per_promoter+=proteins_per_rbs
+            self.made_proteins[rna_construct].update({correct_rbs:proteins_per_rbs})
+            correct_rbs.protein = [a.get_species() for a in proteins_per_rbs]
+            # so this has to be done at the end of the RNA only because only then do we know
+            # exactly what that full RNA is going to be.
+            # we want to clear the current proteins, but there could be multiple RNAs that run over the same proteins,
+            # so wait until we are done tallying all the RNAs before removing everything from current_proteins
+        del self.current_rnas[promoter] # this removes the current RNA from the list, because it's
+                                        # being terminated!!
+        if(self.return_rnas):
+            promoter.transcript = rna_construct.get_species() #set the promoter's transcript to be correct
+        promoter_proteins = [a.get_species() for a in proteins_per_promoter]
+        if(len(promoter_proteins)>0):
+            promoter.protein = promoter_proteins
+        self.all_rnas.update({promoter:rna_construct})
+        return
+    def end(self):
+        """we've reached the end of the dna! End everything!"""
+        logging.debug("end of the DNA!")
+        for promoter in list(self.current_rnas.keys()):
+            self.terminate_transcription(promoter)
+        self.current_rnas = {}
+        self.current_proteins = {}
+        self.current_rbs = None
+    def get_proteins(self):
+        """return all the proteins made during the latest exploration"""
+        return self.made_proteins
+    def get_rnas(self):
+        """return all RNAs made during the latest exploration"""
+        return self.all_rnas
+
+class TxExplorer(TxTlExplorer_CE):
+    def __init__(self,name="TxExplorer",possible_rxns=("transcription",),\
+                        direction="forward",possible_directions=("forward","reverse"),debug=False):
+        TxTlExplorer_CE.__init__(self,name=name,possible_rxns=possible_rxns,\
+                        direction=direction,possible_directions=possible_directions,return_proteins=False)
+
+class TlExplorer(TxTlExplorer_CE):
```

Comment:
I'm a little confused what is going on here - what is TxTlExplorer_CE versus TxExplorer vs TlExplorer?